File: backend/src/main/java/finalproject/backend/service/CourseService.java
package finalproject.backend.service;

import finalproject.backend.modal.CourseLevel;
import finalproject.backend.modal.CourseStatus;
import finalproject.backend.request.CourseRequest;
import finalproject.backend.response.ApiResponse;
import finalproject.backend.response.CourseResponse;
import finalproject.backend.response.LessonResponse;
import finalproject.backend.response.PageResponse;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

public interface CourseService {

    ApiResponse<CourseResponse> createCourse(CourseRequest request, MultipartFile thumbnail);

    PageResponse<CourseResponse> getAllCourses(
            Pageable pageable,
            Integer categoryId,
            CourseStatus status,
            CourseLevel level,
            String search
    );
    ApiResponse<CourseResponse> getCourseById(Long id);

    /** Returns course detail by slug.
     *
     * For page views, this includes nested chapters + lessons in `chapters`.
     */
    ApiResponse<CourseResponse> getCourseBySlug(String slug);

    /** Returns full course with all chapters + lessons nested. */
    ApiResponse<CourseResponse> getCourseWithChapters(String slug);

    /** Returns a single lesson (with code snippets) scoped to a course slug. */
    ApiResponse<LessonResponse> getLessonBySlug(String courseSlug, String lessonSlug);

    PageResponse<CourseResponse> getCoursesByCategory(int categoryId, Pageable pageable);
    PageResponse<CourseResponse> getCoursesByInstructor(Long instructorId, Pageable pageable);

    /** Returns all featured (isFeatured=true) PUBLISHED courses. */
    PageResponse<CourseResponse> getFeaturedCourses(Pageable pageable);

    /** Returns all COMING_SOON courses ordered by launchDate ascending. */
    PageResponse<CourseResponse> getComingSoonCourses(Pageable pageable);

    ApiResponse<CourseResponse> updateCourse(Long id, CourseRequest request, MultipartFile thumbnail);
    ApiResponse<Void> deleteCourse(Long id);
}

File: frontend/src/services/courseService.ts
import { get, post, put, del, buildFormData } from '../lib/BaseApi' // Adjust path to your new functional utilities
import type {
  PageResponse,
  PaginationParams,
  CourseFilterParams,
} from '../types/apiType'

import type {
  LessonResponse
} from '../types/lessonType'

import type { CourseStatus, CourseLevel ,CourseRequest,CourseResponse} from '../types/courseType'


// ═════════════════════════════════════════════════════════════
//  3. COURSE SERVICE
// ═════════════════════════════════════════════════════════════

const COURSE_PATH = '/api/v1/courses'

export const courseService = {
  /** GET / — paginated courses with optional filters */
  getAll: (params: CourseFilterParams = {}): Promise<PageResponse<CourseResponse>> => {
    const { page = 0, size = 10, sortBy = 'createdAt', sortDir = 'desc', categoryId, status, level, search, isFeatured, isFree } = params
    const queryParams: Record<string, unknown> = { page, size, sortBy, sortDir }
    if (categoryId !== undefined) queryParams.categoryId = categoryId
    if (status)                   queryParams.status = status
    if (level)                    queryParams.level = level
    if (search)                   queryParams.search = search
    if (isFeatured !== undefined) queryParams.isFeatured = isFeatured
    // if (isFree !== undefined)     queryParams.isFree = isFree
    return get<PageResponse<CourseResponse>>(COURSE_PATH, { params: queryParams })
  },

  /** GET /:id */
  getById: (id: number): Promise<CourseResponse> => {
    return get<CourseResponse>(`${COURSE_PATH}/${id}`)
  },

  /** GET /slug/:slug */
  getBySlug: (slug: string): Promise<CourseResponse> => {
    return get<CourseResponse>(`${COURSE_PATH}/slug/${slug}`)
  },

  /** GET /slug/:slug/full — course with nested chapters + lessons */
  getWithChapters: (slug: string): Promise<CourseResponse> => {
    return get<CourseResponse>(`${COURSE_PATH}/slug/${slug}/full`)
  },

  /** GET /slug/:courseSlug/lessons/:lessonSlug */
  getLessonBySlug: (courseSlug: string, lessonSlug: string): Promise<LessonResponse> => {
    return get<LessonResponse>(`${COURSE_PATH}/slug/${courseSlug}/lessons/${lessonSlug}`)
  },

  /** GET /category/:categoryId — paginated */
  getByCategory: (
    categoryId: number,
    params: Pick<PaginationParams, 'page' | 'size'> = {}
  ): Promise<PageResponse<CourseResponse>> => {
    const { page = 0, size = 10 } = params
    return get<PageResponse<CourseResponse>>(`${COURSE_PATH}/category/${categoryId}`, { params: { page, size } })
  },

  /** GET /instructor/:instructorId — paginated */
  getByInstructor: (
    instructorId: number,
    params: Pick<PaginationParams, 'page' | 'size'> = {}
  ): Promise<PageResponse<CourseResponse>> => {
    const { page = 0, size = 10 } = params
    return get<PageResponse<CourseResponse>>(`${COURSE_PATH}/instructor/${instructorId}`, { params: { page, size } })
  },

  /** GET /featured — paginated */
  getFeatured: (params: PaginationParams = {}): Promise<PageResponse<CourseResponse>> => {
    const { page = 0, size = 10, sortBy = 'createdAt', sortDir = 'desc' } = params
    return get<PageResponse<CourseResponse>>(`${COURSE_PATH}/featured`, { params: { page, size, sortBy, sortDir } })
  },

  /** GET /coming-soon — paginated, sorted by launchDate asc */
  getComingSoon: (
    params: Pick<PaginationParams, 'page' | 'size'> = {}
  ): Promise<PageResponse<CourseResponse>> => {
    const { page = 0, size = 10 } = params
    return get<PageResponse<CourseResponse>>(`${COURSE_PATH}/coming-soon`, { params: { page, size } })
  },

  /** POST / — [ADMIN] multipart/form-data */
  create: (payload: CourseRequest, thumbnail?: File): Promise<CourseResponse> => {
    const form = buildFormData(payload as unknown as Record<string, unknown>, { thumbnail })
    return post<CourseResponse>(COURSE_PATH, form, { multipart: true })
  },

  /** PUT /:id — [ADMIN] multipart/form-data */
  update: (id: number, payload: CourseRequest, thumbnail?: File): Promise<CourseResponse> => {
    const form = buildFormData(payload as unknown as Record<string, unknown>, { thumbnail })
    return put<CourseResponse>(`${COURSE_PATH}/${id}`, form, { multipart: true })
  },

  /** DELETE /:id — [ADMIN] */
  remove: (id: number): Promise<void> => {
    return del<void>(`${COURSE_PATH}/${id}`, { raw: true })
  }
}
